src/wayland/shell/wlr_layer/handlers.rs

```diff
@@ -308,18 +308,22 @@ where
     }
 
     fn destroyed(
-        _state: &mut D,
+        state: &mut D,
         _client_id: wayland_server::backend::ClientId,
         object_id: wayland_server::backend::ObjectId,
         data: &WlrLayerSurfaceUserData,
     ) {
         data.alive_tracker.destroy_notify();
+
         // remove this surface from the known ones (as well as any leftover dead surface)
-        data.shell_data
-            .known_layers
-            .lock()
-            .unwrap()
-            .retain(|other| other.shell_surface.id() != object_id);
+        let layers = &mut data.shell_data.known_layers.lock().unwrap();
+        if let Some(index) = layers
+            .iter()
+            .position(|layer| layer.shell_surface.id() == object_id)
+        {
+            let layer = layers.remove(index);
+            WlrLayerShellHandler::layer_destroyed(state, layer);
```

Comment:
Is `layers` MutexGuard dropped here by rust automagically? If not, I think you should `mem::drop` it to reduce the possibility of dead locks, not sure where the Mutex is used atm, and you never know where it will be used in future.